huaweicloud/services/ces/resource_huaweicloud_ces_dashboard_widget.go

```diff
@@ -0,0 +1,541 @@
+
+	if _, ok := d.GetOk("unit"); ok {
+		updateWidgetPath := createWidgetClient.Endpoint + "v2/{project_id}/widgets/batch-update"
+		updateWidgetPath = strings.ReplaceAll(updateWidgetPath, "{project_id}", createWidgetClient.ProjectID)
+		updateWidgetOpt := golangsdk.RequestOpts{
+			KeepResponseBody: true,
+			MoreHeaders:      map[string]string{"Content-Type": "application/json;charset=UTF-8"},
+		}
+		param := map[string]interface{}{
+			"widget_id": d.Id(),
+			"unit":      d.Get("unit"),
+		}
+		updateWidgetOpt.JSONBody = []interface{}{param}
+		_, err := createWidgetClient.Request("POST", updateWidgetPath, &updateWidgetOpt)
+		if err != nil {
+			return diag.Errorf("error updating CES dashboard widget: %s", err)
+		}
+	}
+
+	return resourceDashboardWidgetRead(ctx, d, meta)
+}
+
+func buildCreateWidgetBodyParams(d *schema.ResourceData) map[string]interface{} {
+	// Currently, the threshold related functions are invalid.
+	return map[string]interface{}{
+		"title":               d.Get("title"),
+		"metrics":             buildWidgetMetrics(d.Get("metrics")),
+		"view":                d.Get("view"),
+		"metric_display_mode": d.Get("metric_display_mode"),
+		"location":            buildWidgetLocation(d.Get("location")),
+		"properties":          buildWidgetProperties(d.Get("properties")),
+		"threshold_enabled":   false,
+	}
+}
+
+func buildWidgetMetrics(rawParams interface{}) []map[string]interface{} {
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		params := make([]map[string]interface{}, 0, len(rawArray))
+		for _, rawMetric := range rawArray {
+			metric := rawMetric.(map[string]interface{})
+			params = append(params, map[string]interface{}{
+				"namespace":   metric["namespace"],
+				"dimensions":  buildWidgetDimensions(metric["dimensions"]),
+				"metric_name": metric["metric_name"],
+				"alias":       utils.ValueIgnoreEmpty(metric["alias"]),
+			})
+		}
+		return params
+	}
+	return nil
+}
+
+func buildWidgetDimensions(rawParams interface{}) map[string]interface{} {
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		if len(rawArray) == 0 {
+			return nil
+		}
+		raw := rawArray[0].(map[string]interface{})
+		return map[string]interface{}{
+			"name":        raw["name"],
+			"filter_type": raw["filter_type"],
+			"values":      utils.ValueIgnoreEmpty(raw["values"]),
+		}
+	}
+	return nil
+}
+
+func buildWidgetLocation(rawParams interface{}) map[string]interface{} {
+	if rawArray, ok := rawParams.([]interface{}); ok {
+		if len(rawArray) == 0 {
+			return nil
+		}
+		raw := rawArray[0].(map[string]interface{})
+		return map[string]interface{}{
+			"top":    raw["top"],
+			"left":   raw["left"],
+			"width":  raw["width"],
+			"height": raw["height"],
+		}
+	}
+	return nil
+}
+
+func buildWidgetProperties(rawParams interface{}) map[string]interface{} {
+	if rawProperties, ok := rawParams.([]interface{}); ok {
+		if len(rawProperties) == 0 {
+			return nil
+		}
+		raw := rawProperties[0].(map[string]interface{})
```

Comment:
this may cause issue if user only input a properties{}
```suggestion
		raw,ok := rawProperties[0].(map[string]interface{})
		if !ok {
		  continue
		}
```